### PR TITLE
WGAN Porting

### PR DESCRIPTION
**Pull request by shreyasvinaya**
## Description
Porting of WGAN from TF to torch

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/gan.py [hunk 1]
```diff
@@ -0,0 +1,758 @@
+"""Generative Adversarial Networks."""
+import numpy as np
+import torch
+import torch.nn as nn
+import time
+from typing import Callable
+from deepchem.models.torch_models.torch_model import TorchModel
+
+
+class GAN(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/gan.py:10]: Add this class to the docs

File: deepchem/models/torch_models/gan.py [hunk 2]
```diff
@@ -0,0 +1,758 @@
...
+    """Builder class for Generative Adversarial Networks.
+
+    A Generative Adversarial Network (GAN) is a type of generative model. It
+    consists of two parts called the "generator" and the "discriminator". The
+    generator takes random noise as input and transforms it into an output that
+    (hopefully) resembles the training data. The discriminator takes a set of
+    samples as input and tries to distinguish the real training samples from the
+    ones created by the generator. Both of them are trained together. The
+    discriminator tries to get better and better at telling real from false data,
+    while the generator tries to get better and better at fooling the discriminator.
```
Comment (rbharath) [inline deepchem/models/torch_models/gan.py:20]: Add a usage example for this class

File: deepchem/models/torch_models/gan.py [hunk 3]
```diff
@@ -0,0 +1,758 @@
...
+        return self.create_generator_loss(discrim_output_gen)
+
+
+def _list_or_tensor(inputs):
+    if len(inputs) == 1:
+        return inputs[0]
+    return inputs
+
+
+class GANModel(TorchModel):
```
Comment (rbharath) [inline deepchem/models/torch_models/gan.py:343]: Separate this into 2 PRs, the first adding the GAN, the second adding the GANModel

File: deepchem/models/torch_models/gan.py [hunk 4]
```diff
@@ -0,0 +1,758 @@
...
+    Equilibrium in Generative Adversarial Nets (GANs)" (https://arxiv.org/abs/1703.00573).
+    This can lead to better models, and is especially useful for reducing mode
+    collapse, since different generators can learn different parts of the
+    distribution.  To use this technique, simply specify the number of generators
+    and discriminators when calling the constructor.  You can then tell
+    predict_gan_generator() which generator to use for predicting samples.
+
+    Examples
+    --------
+    <examples here>
```
Comment (rbharath) [inline deepchem/models/torch_models/gan.py:397]: Add usage example

File: deepchem/models/torch_models/gan.py [hunk 5]
```diff
@@ -0,0 +1,758 @@
...
+
+    References
+    ----------
+    <references>
```
Comment (rbharath) [inline deepchem/models/torch_models/gan.py:401]: Add reference

File: deepchem/models/torch_models/tests/test_gan.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+                               gradient_penalty=0.1,
+                               model_dir=model_dir)
+    reloaded_gan.restore()
+
+    # See if it has done a plausible job of learning the distribution.
+    means = 10 * np.random.random([1000, 1])
+    batch_size = len(means)
+    noise_input = gan.get_noise_batch(batch_size=batch_size)
+    values = gan.predict_gan_generator(noise_input=noise_input,
+                                       conditional_inputs=[means])
```
Comment (rbharath) [inline deepchem/models/torch_models/tests/test_gan.py:335]: You need unit tests for the GAN layer directly and not just GANModel

**Review by rbharath — COMMENTED** [review]
@shreyasvinaya Please split this into 2 PRs

1. First PR adding GAN class
2. Second PR adding GANModel

The PR is too large for me to do a thorough review in the current format. I will do a more thorough review on the split first PR

**Comment by shreyasvinaya** [comment]
Moved to https://github.com/deepchem/deepchem/pull/3666

**Pull request closed without merging**